### PR TITLE
Adds BLAST+6 format to skbio/io/format.

### PR DESCRIPTION
**Pull request by Kleptobismol**
Adds blast6 module, tests, and test data. Halfway addresses issue #1110.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -3,6 +3,7 @@
 ## Version 0.4.0-dev (changes since 0.4.0 release go here)
 
 ### Features
+* Added ``skbio.io.format.blast6.py`` module as a new format that reads blast6 data to a pandas dataframe. ([#1110](https://github.com/biocore/scikit-bio/issues/1110))
```
Comment (jairideout) [inline CHANGELOG.md]: Can you change this to read:

```
Added `skbio.io.format.blast6` for reading BLAST+ output format 6 or BLAST output format 8 files into a `pd.DataFrame`.
```

I basically want the wording to be as precise as possible since there is often confusion about different BLAST output formats.

File: skbio/io/format/blast6.py [hunk 2]
```diff
@@ -0,0 +1,287 @@
+"""
+BLAST+6 format (:mod:`skbio.io.format.blast6`)
+=========================================================================
```
Comment (jairideout) [inline skbio/io/format/blast6.py]: make this underline match the length of text above it

File: skbio/io/format/blast6.py [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+
+.. currentmodule:: skbio.io.format.blast6
+
+The BLAST+6 format (``blast6``) stores strings, integers, and floats relating
```
Comment (jairideout) [inline skbio/io/format/blast6.py]: use `blast+6` in the parens because that's the string name users will pass

File: skbio/io/format/blast6.py [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+from [1]_::
+
+    moaC    gi|15800534|ref|NP_286546.1|    100.00  161 0   0   1   161 1
+    161 3e-114  330
+    moaC    gi|170768970|ref|ZP_02903423.1| 99.38   161 1   0   1   161 1
+    161 9e-114  329
+
+Format Support
+--------------
+**Has Sniffer: No**
```
Comment (jairideout) [inline skbio/io/format/blast6.py]: Can you add the following line below this line?

```
**State: Experimental as of 0.4.0-dev.**
```

File: skbio/io/format/blast6.py (hunk 3 above)
Comment (jairideout) [inline skbio/io/format/blast6.py]: Can you explain that BLAST+6 format includes BLAST+ output format 6 and BLAST output format 8?

File: skbio/io/format/blast6.py [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+|           |a '<>'                              |     |
++-----------+------------------------------------+-----+
+|qcovs      |Query Coverage Per Subject          |int  |
++-----------+------------------------------------+-----+
+|qcovhsp    |Query Coverage Per HSP              |int  |
++-----------+------------------------------------+-----+
+
+List of data and descriptions taken from BLAST Command Line Applications User
+Manual [2]_.
+
```
Comment (jairideout) [inline skbio/io/format/blast6.py]: Can you add a Format Parameters section here describing `columns` and `default_columns` parameters? See the other I/O formats for examples of how we're formatting this section.

File: skbio/io/format/blast6.py [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+Examples
+--------
+An example of using _blast6_to_data_frame with default column labels.
+
+>>> from io import StringIO
+>>> from skbio import read
+>>> import pandas as pd
+>>> fs = '\\n'.join([
+...     'query1\\tsubject1\\t30.0\\t3\\t5\\t1\\t1\\t3\\t8\\t7\\t0.030\\t13.2',
```
Comment (jairideout) [inline skbio/io/format/blast6.py]: What about using the example file you posted at the top of the docstring here? That one's a little more interesting.

File: skbio/io/format/blast6.py [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+
+List of data and descriptions taken from BLAST Command Line Applications User
+Manual [2]_.
+
+Examples
+--------
+An example of using _blast6_to_data_frame with default column labels.
+
+>>> from io import StringIO
+>>> from skbio import read
```
Comment (jairideout) [inline skbio/io/format/blast6.py]: For this example can you use `import skbio.io`?

File: skbio/io/format/blast6.py [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+
+The BLAST+6 format (``blast6``) stores strings, integers, and floats relating
+to the similarities between multiple DNA sequences. These values are stored in
+a simple tabular format with no labels on individual columns. The values are
+separated by the tab character.
+
+An example BLAST+6-formatted file comparing two protein sequences, taken
+from [1]_::
+
+    moaC    gi|15800534|ref|NP_286546.1|    100.00  161 0   0   1   161 1
```
Comment (jairideout) [inline skbio/io/format/blast6.py]: To emphasize the tab characters, can you insert `<tab>` between the values? These lines also need a line continuation (each line is currently rendering as two lines).

File: skbio/io/format/blast6.py [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
++-----------+------------------------------------+-----+
+|qacc       |Query accesion                      |str  |
++-----------+------------------------------------+-----+
+|qaccver    |Query accesion.version              |str  |
++-----------+------------------------------------+-----+
+|qlen       |Query sequence length               |int  |
++-----------+------------------------------------+-----+
+|sseqid     |Subject Seq-id                      |str  |
++-----------+------------------------------------+-----+
+|sallseqid  |All subject Seq-id(s), separated by\|str  |
```
Comment (jairideout) [inline skbio/io/format/blast6.py]: The second column doesn't render correctly ("bya" instead of "by a"). I think you can remove the line continuation here. There are a few more cells like this below that'll need to be modified as well.

File: skbio/io/format/blast6.py [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
++-----------+------------------------------------+-----+
+|qcovhsp    |Query Coverage Per HSP              |int  |
++-----------+------------------------------------+-----+
+
+List of data and descriptions taken from BLAST Command Line Applications User
+Manual [2]_.
+
+Examples
+--------
+An example of using _blast6_to_data_frame with default column labels.
```
Comment (jairideout) [inline skbio/io/format/blast6.py]: `_blast6_to_data_frame` is hidden from the user

File: skbio/io/format/blast6.py [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+>>> from io import StringIO
+>>> from skbio import read
+>>> import pandas as pd
+>>> fs = '\\n'.join([
+...     'query1\\tsubject1\\t30.0\\t3\\t5\\t1\\t1\\t3\\t8\\t7\\t0.030\\t13.2',
+...     'query2\\tsubject2\\t40.0\\t5\\t4\\t2\\t1\\t6\\t9\\t0\\t0.045\\t15.0'
+... ])
+>>> fh = StringIO(fs)
+
+Load BLAST+6 data into a DataFrame and specify that default columns should be
```
Comment (jairideout) [inline skbio/io/format/blast6.py]: DataFrame -> `pd.DataFrame`

File: skbio/io/format/blast6.py [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+>>> df # doctest: +NORMALIZE_WHITESPACE
+   qseqid    sseqid  pident  length  mismatch  gapopen  qstart  qend  sstart \\
+0  query1  subject1      30       3         5        1       1     3       8
+1  query2  subject2      40       5         4        2       1     6       9
+<BLANKLINE>
+   send  evalue  bitscore
+0     7   0.030      13.2
+1     0   0.045      15.0
+
+An example of using _blast6_to_data_frame with custom column labels.
```
Comment (jairideout) [inline skbio/io/format/blast6.py]: `_blast6_to_data_frame` is hidden from the user

File: skbio/io/format/blast6.py [hunk 13]
```diff
@@ -0,0 +1,287 @@
...
+
+>>> from io import StringIO
+>>> from skbio import read
```
Comment (jairideout) [inline skbio/io/format/blast6.py]: same comment here about using `import skbio.io`

File: skbio/io/format/blast6.py [hunk 14]
```diff
@@ -0,0 +1,287 @@
...
+|salltitles |All Subject Title(s), separated by \|str  |
+|           |a '<>'                              |     |
++-----------+------------------------------------+-----+
+|qcovs      |Query Coverage Per Subject          |int  |
++-----------+------------------------------------+-----+
+|qcovhsp    |Query Coverage Per HSP              |int  |
++-----------+------------------------------------+-----+
+
+List of data and descriptions taken from BLAST Command Line Applications User
+Manual [2]_.
```
Comment (jairideout) [inline skbio/io/format/blast6.py]: Can you add a note below this table describing how N/A values are handled when reading into a DataFrame? i.e. they are converted to `np.nan` (missing data in pandas).

File: skbio/io/format/blast6.py [hunk 15]
```diff
@@ -0,0 +1,287 @@
...
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+from skbio.io import create_format
+import pandas as pd
```
Comment (jairideout) [inline skbio/io/format/blast6.py]: pandas and numpy imports (i.e. 3rd party dependencies) go in their own section before skbio imports

File: skbio/io/format/blast6.py [hunk 16]
```diff
@@ -0,0 +1,287 @@
...
+                     'sframe': int, 'btop': int, 'staxids': str,
+                     'sscinames': str, 'scomnames': str, 'sblastnames': str,
+                     'sskingdoms': str, 'stitle': str, 'salltitles': str,
+                     'sstrand': str, 'qcovs': int, 'qcovhsp': int}
+_default_columns = ['qseqid', 'sseqid', 'pident', 'length', 'mismatch',
+                    'gapopen', 'qstart', 'qend', 'sstart', 'send',
+                    'evalue', 'bitscore']
+
+
+@blast6.reader(pd.DataFrame)
```
Comment (jairideout) [inline skbio/io/format/blast6.py]: pass `monkey_patch=False` here, otherwise `.read()` is added to `pd.DataFrame`. We want to avoid this because it's a class defined external to scikit-bio.

File: skbio/io/format/blast6.py [hunk 17]
```diff
@@ -0,0 +1,287 @@
...
+def _blast6_to_data_frame(fh, columns=None, default_columns=False):
+    if default_columns is True and columns is not None:
```
Comment (jairideout) [inline skbio/io/format/blast6.py]: comparing a boolean to `True` or `False` is unnecessary since it is already a boolean. So this can be simplified to:

```python
if default_columns and columns is not None:
```

same comment for the next `if` statement

File: CHANGELOG.md (hunk 1 above)
Comment (Kleptobismol) [inline CHANGELOG.md]: Done.

File: skbio/io/format/blast6.py (hunk 2 above)
Comment (Kleptobismol) [inline skbio/io/format/blast6.py]: Done.

File: skbio/io/format/blast6.py (hunk 3 above)
Comment (Kleptobismol) [inline skbio/io/format/blast6.py]: Done.

File: skbio/io/format/blast6.py (hunk 4 above)
Comment (Kleptobismol) [inline skbio/io/format/blast6.py]: Done.

File: skbio/io/format/blast6.py [hunk 18]
```diff
@@ -0,0 +1,287 @@
...
+        raise ValueError("You cannot use the default columns and also create"
```
Comment (jairideout) [inline skbio/io/format/blast6.py]: Can you reference the problematic parameters in the error message? Something like:

```
"`columns` and `default_columns` cannot both be provided."
```

File: skbio/io/format/blast6.py [hunk 19]
```diff
@@ -0,0 +1,287 @@
...
+                         " your own columns")
+    if default_columns is False and columns is None:
+        raise ValueError("You must specify whether default columns or custom"
```
Comment (jairideout) [inline skbio/io/format/blast6.py]:
```
"Either `columns` or `default_columns` must be provided."
```

File: skbio/io/format/blast6.py [hunk 20]
```diff
@@ -0,0 +1,287 @@
...
+                         " columns are being used.")
+    if columns is not None:
```
Comment (jairideout) [inline skbio/io/format/blast6.py]: This if/else can be simplified to:

```python
if default_columns:
    columns = _default_columns
```

and then you can use `columns` instead of `names` throughout the rest of the function

File: skbio/io/format/blast6.py [hunk 21]
```diff
@@ -0,0 +1,287 @@
...
+        names = columns
+    else:
+        names = _default_columns
+
+    df = pd.read_csv(fh, na_values='N/A', sep='\t', header=None)
+
+    if len(list(df.columns.values)) != len(names):
```
Comment (jairideout) [inline skbio/io/format/blast6.py]: simplified:

```python
if len(df.columns) != len(names):
```

File: skbio/io/format/blast6.py [hunk 22]
```diff
@@ -0,0 +1,287 @@
...
+        raise ValueError("The amount of data you have supplied does not match"
```
Comment (jairideout) [inline skbio/io/format/blast6.py]: "The amount of data" is very general. Can you make this message say something specific like

```
"The specified number of columns (%d) does not match the number of columns in the file (%d)."
```

Comment (jairideout) [inline skbio/io/format/blast6.py]: Can you also create a new error type `BLAST6FormatError` and raise it here? We've been taking this approach in I/O where file format-specific errors have their own error type. You'll need to add the exception to skbio/io/_exception.py.

I think the other `ValueError`s raised above are okay to leave as-is because they are handling invalid user input not specific to the file format itself.

File: skbio/io/format/blast6.py [hunk 23]
```diff
@@ -0,0 +1,287 @@
...
+                         " the number of columns.")
+
+    try:
+        for i in range(len(names)):
+            df[i] = df[i].astype(_possible_columns[names[i]])
+            df.replace('nan', np.nan, inplace=True)
+        df.columns = names
```
Comment (jairideout) [inline skbio/io/format/blast6.py]: If you set the columns first before looping, the code can be simplified:

```python
df.columns = names
for column in df:
    df[column] = df[column].astype(_possible_columns[column])
```

Comment (jairideout) [inline skbio/io/format/blast6.py]: Can you add an explicit check (and unit test) for the case where the user provides a column that isn't recognized? The current code will just raise a `KeyError` which isn't very informative.

File: skbio/io/format/blast6.py [hunk 24]
```diff
@@ -0,0 +1,287 @@
...
+    df = pd.read_csv(fh, na_values='N/A', sep='\t', header=None)
+
+    if len(list(df.columns.values)) != len(names):
+        raise ValueError("The amount of data you have supplied does not match"
+                         " the number of columns.")
+
+    try:
+        for i in range(len(names)):
+            df[i] = df[i].astype(_possible_columns[names[i]])
+            df.replace('nan', np.nan, inplace=True)
```
Comment (jairideout) [inline skbio/io/format/blast6.py]: Why is this line necessary?

File: skbio/io/format/blast6.py [hunk 25]
```diff
@@ -0,0 +1,287 @@
...
+    else:
+        names = _default_columns
+
+    df = pd.read_csv(fh, na_values='N/A', sep='\t', header=None)
+
+    if len(list(df.columns.values)) != len(names):
+        raise ValueError("The amount of data you have supplied does not match"
+                         " the number of columns.")
+
+    try:
```
Comment (jairideout) [inline skbio/io/format/blast6.py]: Can you move the try/except to wrap the line(s) that could throw this error? I'm guessing it's the `.astype()` line. This will make it clearer where exactly we're trying to catch the pandas error and avoids catching other `ValueErrors` that may be raised by other parts of the code.

File: skbio/io/format/blast6.py (hunk 5 above)
Comment (Kleptobismol) [inline skbio/io/format/blast6.py]: Done.

File: skbio/io/format/tests/test_blast6.py [hunk 26]
```diff
@@ -0,0 +1,95 @@
...
+        fp = get_data_path('blast6_invalid_number_of_columns')
+        with self.assertRaises(ValueError):
+            _blast6_to_data_frame(fp, default_columns=True)
+
+    def test_wrong_column_types_error(self):
+        fp = get_data_path('blast6_invalid_column_types')
+        with self.assertRaises(ValueError):
+            _blast6_to_data_frame(fp, default_columns=True)
+
+    def test_values_under_wrong_column_error(self):
```
Comment (jairideout) [inline skbio/io/format/tests/test_blast6.py]: You can remove this test b/c it's testing the same thing as the test above

File: skbio/io/format/tests/data/blast6_values_under_wrong_column [hunk 27]
```diff
@@ -0,0 +1 @@
+query1	8	75.00	subject2	2	0	1	8	2	0.060	9	11.5
```
Comment (jairideout) [inline skbio/io/format/tests/data/blast6_values_under_wrong_column]: you can remove this file now that its corresponding test is being removed

File: skbio/io/format/tests/test_blast6.py [hunk 28]
```diff
@@ -0,0 +1,95 @@
...
+        fp = get_data_path('blast6_values_under_wrong_column')
+        with self.assertRaises(ValueError):
+            _blast6_to_data_frame(fp, default_columns=True)
+
```
Comment (jairideout) [inline skbio/io/format/tests/test_blast6.py]: Can you also add tests for the case where `default_columns` and `columns` are both omitted, and when both are provided?

Comment (jairideout) [inline skbio/io/format/tests/test_blast6.py]: Can you also add an explicit test for N/A handling? One of your data files already has this but it'd be good to have a separate test for this behavior. I recommend creating a multi-line file using the default columns and adding an N/A or two to each column.

File: skbio/io/format/tests/test_blast6.py [hunk 29]
```diff
@@ -0,0 +1,95 @@
...
+    def test_valid_minimal(self):
+        fp = get_data_path('blast6_custom_minimal')
+        df = _blast6_to_data_frame(fp, columns=['sacc'])
+        exp = pd.DataFrame([['subject2']], columns=['sacc'])
+        assert_data_frame_almost_equal(df, exp)
+
+    def test_wrong_amount_of_columns_error(self):
+        fp = get_data_path('blast6_invalid_number_of_columns')
+        with self.assertRaises(ValueError):
```
Comment (jairideout) [inline skbio/io/format/tests/test_blast6.py]: Can you use `six.assertRaisesRegex` in these tests? There are lots of `ValueErrors` being raised for different reasons, so this is one way to verify the right `ValueError` is being raised in each test. See skbio/sequence/tests/test_sequence.py for examples.

**Comment by jairideout** [comment]
@Kleptobismol thanks for your work on this! I'm done reviewing, let me know when you're ready for another pass.

File: skbio/io/format/blast6.py (hunk 6 above)
Comment (Kleptobismol) [inline skbio/io/format/blast6.py]: Done.

File: skbio/io/format/blast6.py (hunk 7 above)
Comment (Kleptobismol) [inline skbio/io/format/blast6.py]: Done.

File: skbio/io/format/blast6.py (hunk 8 above)
Comment (Kleptobismol) [inline skbio/io/format/blast6.py]: Done.

File: skbio/io/format/blast6.py (hunk 9 above)
Comment (Kleptobismol) [inline skbio/io/format/blast6.py]: Let's talk about this one.

File: skbio/io/format/blast6.py (hunk 10 above)
Comment (Kleptobismol) [inline skbio/io/format/blast6.py]: Done.

File: skbio/io/format/blast6.py (hunk 11 above)
Comment (Kleptobismol) [inline skbio/io/format/blast6.py]: Done.

File: skbio/io/format/blast6.py (hunk 12 above)
Comment (Kleptobismol) [inline skbio/io/format/blast6.py]: Done.

File: skbio/io/format/blast6.py (hunk 13 above)
Comment (Kleptobismol) [inline skbio/io/format/blast6.py]: Done.

File: skbio/io/format/blast6.py (hunk 15 above)
Comment (Kleptobismol) [inline skbio/io/format/blast6.py]: Done.

File: skbio/io/format/blast6.py (hunk 16 above)
Comment (Kleptobismol) [inline skbio/io/format/blast6.py]: Done.

File: skbio/io/format/blast6.py (hunk 17 above)
Comment (Kleptobismol) [inline skbio/io/format/blast6.py]: Done.

File: skbio/io/format/blast6.py (hunk 18 above)
Comment (Kleptobismol) [inline skbio/io/format/blast6.py]: Done.

File: skbio/io/format/blast6.py (hunk 19 above)
Comment (Kleptobismol) [inline skbio/io/format/blast6.py]: Done.

File: skbio/io/format/blast6.py (hunk 20 above)
Comment (Kleptobismol) [inline skbio/io/format/blast6.py]: Done.

File: skbio/io/format/blast6.py (hunk 21 above)
Comment (Kleptobismol) [inline skbio/io/format/blast6.py]: Done.

File: skbio/io/format/blast6.py (hunk 22 above)
Comment (Kleptobismol) [inline skbio/io/format/blast6.py]: Done.

File: skbio/io/format/blast6.py (hunk 14 above)
Comment (Kleptobismol) [inline skbio/io/format/blast6.py]: Done.

File: skbio/io/format/blast6.py (hunk 23 above)
Comment (Kleptobismol) [inline skbio/io/format/blast6.py]: Done.

File: skbio/io/format/tests/test_blast6.py (hunk 26 above)
Comment (Kleptobismol) [inline skbio/io/format/tests/test_blast6.py]: Done.

File: skbio/io/format/tests/data/blast6_values_under_wrong_column (hunk 27 above)
Comment (Kleptobismol) [inline skbio/io/format/tests/data/blast6_values_under_wrong_column]: Done.

File: skbio/io/format/tests/test_blast6.py (hunk 29 above)
Comment (Kleptobismol) [inline skbio/io/format/tests/test_blast6.py]: Done.

File: skbio/io/format/tests/test_blast6.py (hunk 28 above)
Comment (Kleptobismol) [inline skbio/io/format/tests/test_blast6.py]: Done.

File: skbio/io/format/blast6.py (hunk 9 above)
Comment (Kleptobismol) [inline skbio/io/format/blast6.py]: Done.

**Comment by Kleptobismol** [comment]
Sorry @gregcaporaso for asking this question in my PR message, my bad. 
I noticed that N/A values seem to only exist in ``str`` type columns. Also my code throws an error when N/A values appear in an ``int`` or ``float`` type column. I was just wondering if you've ever seen BLAST return N/A values in any of these columns before:
``qgi`` ``qlen`` ``sgi`` ``sallgi`` ``slen`` ``qstart`` ``qend`` ``sstart`` ``send`` ``evalue`` ``bitscore`` ``score`` 
``length`` ``pident`` ``nident`` ``mismatch`` ``positive`` ``gapopen`` ``gaps`` ``ppos`` ``qframe`` ``sframe``
``btop`` ``qcovs`` ``qcovshsp``

File: skbio/io/format/tests/data/blast6_values_under_wrong_column (hunk 27 above)
Comment (jairideout) [inline skbio/io/format/tests/data/blast6_values_under_wrong_column]: This file is still tracked, you'll need to remove it with `git rm`, commit, and push.

File: skbio/io/format/blast6.py [hunk 30]
```diff
@@ -0,0 +1,311 @@
...
+    if default_columns and columns is not None:
+        raise ValueError("`columns` and `default_columns` cannot both be"
+                         " provided.")
+    if not default_columns and columns is None:
+        raise ValueError("Either `columns` or `default_columns` must be"
+                         " provided.")
+    if default_columns:
+        columns = _default_columns
+
+    df = pd.read_csv(fh, na_values='N/A', sep='\t', header=None)
```
Comment (jairideout) [inline skbio/io/format/blast6.py]: add `keep_default_na=False` to avoid pandas recognizing other missing data representations (we only want "N/A" to be considered missing)

File: skbio/io/format/tests/test_blast6.py [hunk 31]
```diff
@@ -0,0 +1,123 @@
...
+
+    def test_different_data_in_same_column(self):
+        fp = get_data_path('blast6_invalid_type_in_column')
+        with self.assertRaises(ValueError):
+            _blast6_to_data_frame(fp, default_columns=True)
+
+    def test_wrong_column_name_error(self):
+        fp = get_data_path('blast6_default_single_line')
+        with assertRaisesRegex(self, ValueError,
+                               "Unrecognized column 'abcd'"):
```
Comment (jairideout) [inline skbio/io/format/tests/test_blast6.py]: This regex is failing in Python 2 because it's repr'd as `u'abcd'`. Modifying the regex to be `Unrecognized column.*'abcd'` should make this work in Python 2 and 3.

**Comment by jairideout** [comment]
Thanks @Kleptobismol, this looks great! @ebolyen, I/O wizard, can you please take a quick pass through and merge if you think it's good to go?

**Comment by ebolyen** [comment]
Thanks @Kleptobismol and @jairideout!
